File: db4o.cs/src/com/db4o/cs/internal/messages/MsgQuery.java
/* Copyright (C) 2006  Versant Inc.  http://www.db4o.com */

package com.db4o.cs.internal.messages;

import com.db4o.config.*;
import com.db4o.cs.internal.*;
import com.db4o.cs.internal.objectexchange.*;
import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.query.result.*;

public abstract class MsgQuery extends MsgObject {
	
	private static int nextID;
	
	protected final MsgD writeQueryResult(AbstractQueryResult queryResult, QueryEvaluationMode evaluationMode, ObjectExchangeConfiguration config) {
		
		if(evaluationMode == QueryEvaluationMode.IMMEDIATE){
			return writeImmediateQueryResult(queryResult, config);
		} 
		return writeLazyQueryResult(queryResult, config);
	}

	private MsgD writeLazyQueryResult(AbstractQueryResult queryResult, ObjectExchangeConfiguration config) {
	    int queryResultId = generateID();
	    int maxCount = config().prefetchObjectCount();
	    IntIterator4 idIterator = queryResult.iterateIDs();
	    MsgD message = buildQueryResultMessage(queryResultId, idIterator, maxCount, config);
	    ServerMessageDispatcher serverThread = serverMessageDispatcher();
	    serverThread.mapQueryResultToID(new LazyClientObjectSetStub(queryResult, idIterator), queryResultId);
	    return message;
    }

	private MsgD writeImmediateQueryResult(AbstractQueryResult queryResult, ObjectExchangeConfiguration config) {
	    IntIterator4 idIterator = queryResult.iterateIDs();
	    MsgD message = buildQueryResultMessage(0, idIterator, queryResult.size(), config);
	    return message;
    }

	private MsgD buildQueryResultMessage(int queryResultId, IntIterator4 ids, int maxCount, ObjectExchangeConfiguration config) {
		final ByteArrayBuffer payload = ObjectExchangeStrategyFactory.forConfig(config).marshall((LocalTransaction) transaction(), ids, maxCount);
	    MsgD message = QUERY_RESULT.getWriterForLength(transaction(), Const4.INT_LENGTH + payload.length());
		StatefulBuffer writer = message.payLoad();
		writer.writeInt(queryResultId);
		writer.writeBytes(payload._buffer);
	    return message;
    }
	
	private static synchronized int generateID(){
		nextID ++;
		if(nextID < 0){
			nextID = 1;
		}
		return nextID;
	}
	
	protected AbstractQueryResult newQueryResult(QueryEvaluationMode mode){
		return container().newQueryResult(transaction(), mode);
	}

}

File: db4o.net/Db4objects.Db4o.CS/Db4objects.Db4o.CS/CS/Internal/Messages/MsgQuery.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4objects.Db4o.CS.Internal;
using Db4objects.Db4o.CS.Internal.Messages;
using Db4objects.Db4o.CS.Internal.Objectexchange;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Query.Result;

namespace Db4objects.Db4o.CS.Internal.Messages
{
	public abstract class MsgQuery : MsgObject
	{
		private static int nextID;

		protected MsgD WriteQueryResult(AbstractQueryResult queryResult, QueryEvaluationMode
			 evaluationMode, ObjectExchangeConfiguration config)
		{
			if (evaluationMode == QueryEvaluationMode.Immediate)
			{
				return WriteImmediateQueryResult(queryResult, config);
			}
			return WriteLazyQueryResult(queryResult, config);
		}

		private MsgD WriteLazyQueryResult(AbstractQueryResult queryResult, ObjectExchangeConfiguration
			 config)
		{
			int queryResultId = GenerateID();
			int maxCount = Config().PrefetchObjectCount();
			IIntIterator4 idIterator = queryResult.IterateIDs();
			MsgD message = BuildQueryResultMessage(queryResultId, idIterator, maxCount, config
				);
			IServerMessageDispatcher serverThread = ServerMessageDispatcher();
			serverThread.MapQueryResultToID(new LazyClientObjectSetStub(queryResult, idIterator
				), queryResultId);
			return message;
		}

		private MsgD WriteImmediateQueryResult(AbstractQueryResult queryResult, ObjectExchangeConfiguration
			 config)
		{
			IIntIterator4 idIterator = queryResult.IterateIDs();
			MsgD message = BuildQueryResultMessage(0, idIterator, queryResult.Size(), config);
			return message;
		}

		private MsgD BuildQueryResultMessage(int queryResultId, IIntIterator4 ids, int maxCount
			, ObjectExchangeConfiguration config)
		{
			ByteArrayBuffer payload = ObjectExchangeStrategyFactory.ForConfig(config).Marshall
				((LocalTransaction)Transaction(), ids, maxCount);
			MsgD message = QueryResult.GetWriterForLength(Transaction(), Const4.IntLength + payload
				.Length());
			StatefulBuffer writer = message.PayLoad();
			writer.WriteInt(queryResultId);
			writer.WriteBytes(payload._buffer);
			return message;
		}

		private static int GenerateID()
		{
			lock (typeof(MsgQuery))
			{
				nextID++;
				if (nextID < 0)
				{
					nextID = 1;
				}
				return nextID;
			}
		}

		protected virtual AbstractQueryResult NewQueryResult(QueryEvaluationMode mode)
		{
			return Container().NewQueryResult(Transaction(), mode);
		}
	}
}
